### ISSUE TITLE
Automatically run and track arrow / parquet benchmarks over time

### ISSUE THREAD

**Issue by alamb** [apache/arrow-rs#1274]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
As discussed by @jonkeane in https://github.com/apache/arrow-site/pull/188#discussion_r799558584

It would be nice to track the speed of rust benchmarks over time to
1. Ensure it doesn't regress (get worse) accidentally
2. We can see the results of improvements over time (and link to them publicly)

**Describe the solution you'd like**
Figure out how to run the benchmarks `cargo bench ...` in the framework described here: https://lists.apache.org/thread/n98tr38moqvbf1rlqd7n7912mfrqv8ps

**Describe alternatives you've considered**

**Additional context**

**Comment by kazuk** [comment]
I write bash script for this.

usage example `./dev/bench/run_bench.sh 11.0.0 12.0.0`

This saves 4 baseline: arrow-11.0.0 and arrow-12.0.0 , parquet-11.0.0, parquet-12.0.0,

this can compare with [critcmp](https://github.com/BurntSushi/critcmp)

```bash
#!/bin/bash
#
# run bench for 2 git hash/branch/tag
#
# ./dev/bench/run_bench.sh [baseline] [compare]

set -e

function patch_cargo_toml() {
    if [ ! -z $(grep "bench = false" arrow/Cargo.toml) ]; then 
        echo "Patch not required"
    else
        echo "Patch required"
        cp arrow/Cargo.toml arrow/Cargo.toml.org
        sed '/^\[lib\]/a bench = false' arrow/Cargo.toml.org > arrow/Cargo.toml
        echo '[lib]' >> parquet/Cargo.toml
        echo 'bench = false' >> parquet/Cargo.toml
    fi
}

function run_bench( ) {
    cd arrow
    cargo bench -- --save-baseline arrow-$1
    cd ..
    cd parquet
    cargo bench -- --save-baseline parquet-$1
    cd ..
}


# checkout baseline and run bench 
git checkout $1

patch_cargo_toml 
run_bench $1

# checkout compare and run bench
git reset --hard
git checkout $2

patch_cargo_toml 
run_bench $2
```

**Comment by kazuk** [comment]
Why critcmp use?

criterion's `--save-baseline` and `--baseline` crashes new bench added in base..compare history.

this is fixed in https://github.com/bheisler/criterion.rs/pull/532 but not released.

**Comment by rok** [comment]
Hey. I'm currently migrating arrow-cpp's benchmarking infrastructure from voltron's AWS account to Arrow project's AWS account. Is there still need for benchmarking on arrow-rs?

**Comment by alamb** [comment]
> Hey. I'm currently migrating arrow-cpp's benchmarking infrastructure from voltron's AWS account to Arrow project's AWS account. Is there still need for benchmarking on arrow-rs?

Yes for sure, thanks @rok ! -- I don't have the bandwidth to really drive the infrastructure part (e.g. manage instances, trigger, jobs, etc) but I would love to help get arrow-rs benchmark results into the infrastructure.

Basically if you can give me some hints about what format the benchmarks need to produce results in I help write some conversion logic from the criterion benchmarks to that other fomat

**Comment by rok** [comment]
Great to know @alamb. I'll be able to give a good answer in a couple of days as I get to know our setup a little better.

**Comment by rok** [comment]
Sorry for low activity @alamb, can't spare the much time right now.
Do you think [arrow-go-like](https://github.com/apache/arrow-go/blob/main/ci/scripts/bench.sh) approach would work for arrow-rust or should we provide ec2 machines?

**Comment by alamb** [comment]
FYI what I have done is to make a bot on a VM that I run (and InfluxData pays for) 

You can see it at work here: https://github.com/apache/arrow-rs/pull/9236#issuecomment-3777725891

It is pretty grotty and doesn't track performance over time but has seemingly really accelerated performance optimizations (as it is much easier to run benchmarks)

**Comment by rok** [comment]
That sounds great!
By the way conbench has [benchclient](https://conbench.github.io/conbench/) to send benchmark results to conbench database. If you'd like we can set it up.

**Comment by Rich-T-kid** [comment]
This seems pretty straightforward to implement, I can start working on this.

I guess this most important piece of this issue is determining what benchmarks we'd like to use as a guage betweeen versions. In datafusion this is an easier question to answer since there are standards already out (clickbench,TPCH/10...ect).

Id reccomend 
1. filter kernels (`take`,`filter_record_batch`)
2. BatchCoalescer kernels (RepartionExec)
3. concatenate kernels 
4. row format kernels  (Aggregations/sorts)
5. sort kernel 
6. cast kernel
7. ipc kernel  (flushing to disk)

due to their direct ties to query exection in datafusion downstream


then we'd also have the parquet related kernels 

1. parquet reader kernel
2. parquet writer kernel

Im personally not super familiar with all the sub-sections of the parquet kernel so any other reqs would be useful.

It may also be nice to include the side by side differneces between major releases as well, similar to what @alamb does for [datafusion major releases ](https://datafusion.apache.org/blog/output/2026/08/25/datafusion-55.0.0/#appendix-clickbench-results).


I'll let this sit for a few days in case anyone has opinions on which kenerls to include/leave out then ill open a PR 🚀

**Comment by Rich-T-kid** [comment]
take

**Comment by sdf-jkl** [comment]
@Rich-T-kid if we move forward with the [codspeed PR](https://github.com/apache/arrow-rs/pull/9975) it should do the tracking for us.

**Comment by Rich-T-kid** [comment]
@sdf-jkl nice

**Comment by Rich-T-kid** [comment]
is there a way to un-take a issue?

**Comment by Rich-T-kid** [comment]
untake
